### ISSUE TITLE
Request scoped query filters for multi-tenant applications

### ISSUE THREAD

**Issue by klausXR** [knex/knex#4823]
I will rephrase this to be a feature request with an API proposition, but I couldn't figure out how to do it from the docs, so I assume that it may be an existing functionality.

I'm working on a multi-tenant application that accomplishes multi-tenancy through a `tenant_id` column in every table. This poses some security risks though, with database access/writes, because the `tenant_id` must always be correct and manually passed.

I'm wondering if there's a way to create a request scoped knex instance, that will be used across the code, and will automatically inject the `tenant_id` into every sql query ( read or write ) to the database, if the table is marked as multi-tenant one.

Here's a short example

```js
app.use((req, res) => {
    // The `req.tenant` is populated by another middleware, we just
    // scope it to the knex instance
    req.knex = knex.createScope({
        tenant_id: req.tenant.id
    })
})

app.get("/user/:id", async (req, res) => {
    // Use the knex instance from the request
    // `tenant_id` should be auto-appended to the query
    const user = await req.knex("users")
        .where({ id: req.params.id })

    res.send(user)
})

app.post("/user", async (req, res) => {
    // Use the knex instance from the request
    // `tenant_id` should be auto-appended to the write op
    const user = await req.knex("users")
        .insert({
            username: req.body.username,
            email: req.body.email
        })

    res.send(user)
})
```

Is this achievable currently through some of the API's or would it be a feature request?

**Comment by kibertoad** [comment]
I don't think this is currently possible. PR would be welcome.

**Comment by klausXR** [comment]
Okay. I don't really have the time at the moment, but the lack of expertise would be the bigger problem. Would you assist me with the terminology here, if this approach already has a conventional name, so that I could make the issue more discoverable for other people searching for it and see if it draws any attention. 

I can at least help with the API, and if you could point me to a module in the code where a feature like that can be implemented, I will be sure to check it out when I have some time available, to determine whether or not it is something that I can technically accomplish. 

With SaaS becoming more and more widespread, I think this way of managing multi-tenancy is one of the more common ones, so this feature should be useful in theory to a lot of people that chose knex over an orm.

**Comment by kibertoad** [comment]
My suggestion would be to use awilix to do per-request injection (it has wrappers for express and fastify) but inject not knex per se, but a function which on invocation returns knex builder with that parameter already set (you can't reuse it across queries tho, would need to invoke that function every time).
Happy to elaborate if needed.

**Comment by elhigu** [comment]
I don't think that this kind of feature should be implemented on query builder level. If they bring enough these ormy feaures to knex, it will eventually become an ORM (that was actually how objection was initiated).

With objection.js in the other hand this is pretty trivial to implement, you don't have to call it ORM if you prefer not to use an ORM (objectio.js does not add any limitations what kind of queries can be made and it does not add any additional magic if you choose to not use those magical parts like relations ).

**Comment by klausXR** [comment]
I wasn't really asking for a feature designed to do this, but an API that would allow one to accomplish this, if so desired.

If we take this example from the docs for the `start` event

```js
knex.select('*')
  .from('users')
  .on('start', function(builder) {
    builder
    .where('IsPrivate', 0)
  })
  .then(function(Rows) {
    //Only contains Rows where IsPrivate = 0
  })
  .catch(function(error) { });
  ```
  
  If there was a hook that was fired before a `select` or `update` clause, I would be able to add a custom hook ( knex already supports scoped contexts, as far as I can see for the naming convention thing ) with the `tenant_id` and add it manually.
  
 So, I by no means want an explicit feature for multi-tenancy, but possibly an API that would allow me to accomplish it on a global level, like create a new instance of Knex and add hooks to it that  apply to this specific instance only.

**Comment by elhigu** [comment]
Having select / update hooks is not as trivial as it sounds, one would need to define and test how to deal with sub queries, CTE stuff, raw queries and all that... not impossible, but not trivial either. 

Anyways you can hook to the common hook that is fired before query being compiled and then check what kind of query is being made and modify it if needed. That way it is very generic and you may even be able to use that knex feature to create those other hooks that you would like to have those.

I would like to keep knex simple and powerful and make all more complicated and possibly opinionated features to be implemented as libraries / helpers / recipes etc.

**Comment by elhigu** [comment]
Event handlers can be also set in global level in knex so you won't need to add them to every query separately.

**Comment by klausXR** [comment]
As long as there can be scoped context added to a knex instance, its ok if the hooks can be global, because the context will be different for every hook. 

I will see how the compilation works exactly and figure out if I can accomplish it with the current API and what edge cases occur and I will keep this issue updated, if anybody else stumbles upon it.

Thank you both for the quick responses and the insights. You have done amazing work with this library and there are a few projects built on top of it already ( at least Objection and MikroOrm ).

**Comment by harish2704** [comment]
IMHO, more general purpose solution is  to implement a global "beforeQuery" hook . 
In that case, we will have more flexibility regarding "what to do" & "when to do" depending on query context. 
Also, in this way, knex will not loose its simplicity

**Comment by mercmobily** [comment]
Excellent news for multi-tenant applications! This feature is being implemented through our new connection context system.

We're adding connection contexts (#6260) that allow you to pin multiple queries to the same connection without requiring a transaction. This enables exactly the request-scoped query filters you need for multi-tenant applications.

This is part of a comprehensive effort (#6257) to modernise Knex's transaction and connection management.

The new API will support request-scoped tenant isolation:
```javascript
// Middleware to set up tenant context
app.use(async (req, res, next) => {
  req.ctx = knex.context(); // Pin to single connection
  
  // Set tenant context at connection level
  await req.ctx.raw('SET SESSION application.tenant_id = ?', [req.tenantId]);
  // Or for Postgres RLS
  await req.ctx.raw('SET LOCAL app.current_tenant = ?', [req.tenantId]);
  
  next();
});

// All queries in this request use the same connection with tenant context
app.get('/api/orders', async (req, res) => {
  // These queries all run on the same connection with tenant isolation
  const orders = await req.ctx('orders').select();
  const items = await req.ctx('order_items').whereIn('order_id', orderIds);
  
  req.ctx.end(); // Release connection
  res.json({ orders, items });
});
```

This provides the request-scoped connection management needed for proper multi-tenant isolation. Thank you for your patience with this important feature request. Closing in favour of the implementation issue.
